### PR TITLE
PYIC-3114 Added new cimit stub base proj. for cimit stub lambda's, added getContraIndicators lamdba with test.

### PR DESCRIPTION
**Pull request by rpayal**
## Proposed changes
Create initial `getContraIndicators` stub lambda

### What changed
- Added new cimit stub base proj. for cimit stub lambda's
- Added getContraIndicators lamdba with test.

### Issue tracking

- [PYI-3114](https://govukverify.atlassian.net/browse/PYI-3114)

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: .github/workflows/deploy-cimit-stub.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+          role-to-assume: ${{ secrets.CORE_CIMIT_GH_ACTIONS_ROLE_ARN }}
+          aws-region: eu-west-2
+
+      - name: Set up SAM cli
+        uses: aws-actions/setup-sam@v2
+
+      - name: Set up esbuild
+        run: curl -fsSL https://esbuild.github.io/dl/latest | sh
+
+      - name: add esbuild to path
```
Comment (therealslimsladey) [inline .github/workflows/deploy-cimit-stub.yml]: we won't need esbuild?

File: .github/workflows/deploy-cimit-stub.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
+name: CIMIT Stub Build & Push
```
Comment (therealslimsladey) [inline .github/workflows/deploy-cimit-stub.yml]: As discussed, can we move this to a separate PR for separate review by the SREs

File: di-ipv-cimit-stub/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+### CIMIT lambda stubs
+This will create stub lambda's for CIMIT lambda's getContraIndicators, getContraIndicatorCredential,
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/README.md]: Amend to 'This will create stubs for the CIMIT lambdas `getContraIndicators` ...

File: di-ipv-cimit-stub/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+putContraIndicators and postMitigations
+The format of the postMitigations POST request should look like
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/README.md]: Maybe change to 
sample `postMitigations` request
and similarly on the other examples

File: di-ipv-cimit-stub/deploy/template.yaml [hunk 5]
```diff
@@ -0,0 +1,115 @@
+AWSTemplateFormatVersion: "2010-09-09"
+
+Transform: AWS::Serverless-2016-10-31
+Globals:
+  Function:
+    Timeout: 30
+
+Description: >-
+  This creates the lambda function to stub CIMIT lambda's for getContraIndicators, getContraIndicatorCredential,
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/deploy/template.yaml]: function -> functions, lambda's -> lambdas

File: di-ipv-cimit-stub/deploy/template.yaml [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+      The name of the stack that defines the VPC in which this container will
+      run.
+    Type: String
+  Environment:
+    Description: The name of the environment to deploy to.
+    Type: String
+    AllowedPattern: ((production)|(build)|(dev.*))
+
+Resources:
+  # lambda to replicate cimit getContraIndicators
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/deploy/template.yaml]: lambda to stub CIMIT getContraIndicators

File: di-ipv-cimit-stub/deploy/template.yaml [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+          - !GetAtt CimitLambdaSecurityGroup.GroupId
+      Policies:
+        - VPCAccessPolicy: { }
+        - Statement:
+            - Sid: AllowGetSSMParam
+              Effect: Allow
+              Action:
+                - "ssm:GetParameter"
+              Resource:
+                - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/stubs/*"
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/deploy/template.yaml]: There's a Policy template we can use, `SSMParameterReadPolicy`, for example,
```
        - SSMParameterReadPolicy:
            ParameterName: !Sub ${Environment}/stubs/cimit/*
```

File: di-ipv-cimit-stub/deploy/template.yaml [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+              - "kms:Encrypt*"
+              - "kms:Decrypt*"
+              - "kms:ReEncrypt*"
+              - "kms:GenerateDataKey*"
+              - "kms:Describe*"
+            Resource: "*"
+            Condition:
+              ArnLike:
+                "kms:EncryptionContext:aws:logs:arn": !Sub "arn:aws:logs:${AWS::Region}:${AWS::AccountId}:*"
+
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/deploy/template.yaml]: extraneous blank lines

File: di-ipv-cimit-stub/deploy/template.yaml [hunk 9]
```diff
@@ -0,0 +1,115 @@
+AWSTemplateFormatVersion: "2010-09-09"
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/deploy/template.yaml:1]: we should get SRE eyes on this

File: di-ipv-cimit-stub/gradle/wrapper/gradle-wrapper.jar
Comment (therealslimsladey) [inline di-ipv-cimit-stub/gradle/wrapper/gradle-wrapper.jar:1]: I think we should use same version of gradle as rest of project and bump them all at same time

File: di-ipv-cimit-stub/lambdas/get-contra-indicators/build.gradle [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+repositories {
+	mavenCentral()
+}
+
+sourceCompatibility = 17
+
+dependencies {
+	implementation "com.amazonaws:aws-lambda-java-core:1.2.1",
+			"com.amazonaws:aws-lambda-java-events:3.11.0",
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/lambdas/get-contra-indicators/build.gradle]: Do we need this dependency?

File: di-ipv-cimit-stub/lambdas/get-contra-indicators/src/main/java/uk/gov/di/ipv/core/getcontraindicators/GetContraIndicatorsHandler.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.util.Collections;
+
+public class GetContraIndicatorsHandler implements RequestHandler<GetCiRequest, GetCiResponse> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GetContraIndicatorsHandler.class);
+
+    @Override
+    public GetCiResponse handleRequest(GetCiRequest event, Context context) {
+        LOGGER.info("EVENT TYPE: " + event.getClass().toString());
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/lambdas/get-contra-indicators/src/main/java/uk/gov/di/ipv/core/getcontraindicators/GetContraIndicatorsHandler.java]: Use StringMappMessage?

File: di-ipv-cimit-stub/lambdas/get-contra-indicators/src/main/java/uk/gov/di/ipv/core/getcontraindicators/domain/ContraIndicatorItem.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
+package uk.gov.di.ipv.core.getcontraindicators.domain;
+
+import lombok.Data;
+
+import java.time.Instant;
+import java.time.format.DateTimeFormatter;
+
+@Data
+public class ContraIndicatorItem implements Comparable<ContraIndicatorItem> {
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/lambdas/get-contra-indicators/src/main/java/uk/gov/di/ipv/core/getcontraindicators/domain/ContraIndicatorItem.java]: ContraIndicatorItem in the stub won't need to be Comparable

File: di-ipv-cimit-stub/lambdas/get-contra-indicators/src/main/java/uk/gov/di/ipv/core/getcontraindicators/domain/ContraIndicatorItem.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    private final String userId;
+    private final String sortKey;
+    private final String iss;
+    private final String issuedAt;
+    private final String ci;
+    private final String ttl;
+    private final String documentId;
+
+    @Override
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/lambdas/get-contra-indicators/src/main/java/uk/gov/di/ipv/core/getcontraindicators/domain/ContraIndicatorItem.java]: Can lose this when the class no longer comparable

File: settings.gradle [hunk 14]
```diff
@@ -2,4 +2,7 @@ include "di-ipv-core-stub",
 		"di-ipv-credential-issuer-stub",
 		"di-ipv-orchestrator-stub",
 		"experian-kbv-stub",
-		"experian-fraud-stub"
+		"experian-fraud-stub",
```
Comment (therealslimsladey) [inline settings.gradle:5]: Do we need all 3 levels?

**Review by therealslimsladey — CHANGES_REQUESTED** [review]

File: di-ipv-cimit-stub/deploy/template.yaml [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+      LogGroupName: !Sub "/aws/lambda/get-contra-indicators-${Environment}"
+      KmsKeyId: !GetAtt LoggingKmsKey.Arn
+
+  CimitLambdaSecurityGroup:
+    Type: "AWS::EC2::SecurityGroup"
+    Properties:
+      GroupDescription: >-
+        CIMIT Lambda Security Group
+      SecurityGroupEgress:
+        - CidrIp: 0.0.0.0/0
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/deploy/template.yaml]: Please remove the Egress rule - the stubs shouldn't need internet access

File: di-ipv-cimit-stub/deploy/template.yaml (hunk 5 above)
Comment (rpayal) [reply di-ipv-cimit-stub/deploy/template.yaml]: done

File: di-ipv-cimit-stub/README.md (hunk 4 above)
Comment (rpayal) [reply di-ipv-cimit-stub/README.md]: Done

File: di-ipv-cimit-stub/README.md (hunk 3 above)
Comment (rpayal) [reply di-ipv-cimit-stub/README.md]: Done

File: di-ipv-cimit-stub/lambdas/get-contra-indicators/src/main/java/uk/gov/di/ipv/core/getcontraindicators/domain/ContraIndicatorItem.java (hunk 12 above)
Comment (rpayal) [reply di-ipv-cimit-stub/lambdas/get-contra-indicators/src/main/java/uk/gov/di/ipv/core/getcontraindicators/domain/ContraIndicatorItem.java]: done

File: di-ipv-cimit-stub/lambdas/get-contra-indicators/src/main/java/uk/gov/di/ipv/core/getcontraindicators/domain/ContraIndicatorItem.java (hunk 13 above)
Comment (rpayal) [reply di-ipv-cimit-stub/lambdas/get-contra-indicators/src/main/java/uk/gov/di/ipv/core/getcontraindicators/domain/ContraIndicatorItem.java]: done

File: settings.gradle (hunk 14 above)
Comment (rpayal) [reply settings.gradle:5]: done

File: di-ipv-cimit-stub/lambdas/get-contra-indicators/build.gradle (hunk 10 above)
Comment (rpayal) [reply di-ipv-cimit-stub/lambdas/get-contra-indicators/build.gradle]: done

File: di-ipv-cimit-stub/lambdas/get-contra-indicators/src/main/java/uk/gov/di/ipv/core/getcontraindicators/GetContraIndicatorsHandler.java (hunk 11 above)
Comment (rpayal) [reply di-ipv-cimit-stub/lambdas/get-contra-indicators/src/main/java/uk/gov/di/ipv/core/getcontraindicators/GetContraIndicatorsHandler.java]: done

File: di-ipv-cimit-stub/deploy/template.yaml (hunk 15 above)
Comment (rpayal) [reply di-ipv-cimit-stub/deploy/template.yaml]: done

**Review by therealslimsladey — APPROVED** [review]
There are a few minor comments left unresolved. Up to you whether we resolve them now or in subsequent PR

File: di-ipv-cimit-stub/deploy/template.yaml [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+              Action:
+                - "ssm:GetParameter"
+              Resource:
+                - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/stubs/*"
+
+  GetContraIndicatorsFunctionLogGroup:
+    Type: AWS::Logs::LogGroup
+    Properties:
+      RetentionInDays: 14
+      LogGroupName: !Sub "/aws/lambda/get-contra-indicators-${Environment}"
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/deploy/template.yaml]: The log group name needs to be named as per the lambda, so
`!Sub "/aws/lambda/getContraIndicators-${Environment}"`
Currently we're getting an empty log group created by this template and a log group named per the lambda that gets autocreated

**Review by therealslimsladey — CHANGES_REQUESTED** [review]
The log group naming in the SAM template needs correcting

File: di-ipv-cimit-stub/deploy/template.yaml [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+      run.
+    Type: String
+  Environment:
+    Description: The name of the environment to deploy to.
+    Type: String
+    AllowedPattern: ((production)|(build)|(dev.*))
+
+Resources:
+  # lambda to replicate cimit getContraIndicators
+  GetContraIndicatorsLambda:
```
Comment (therealslimsladey) [inline di-ipv-cimit-stub/deploy/template.yaml]: Rename resource as `GetContraIndicatorsFunction` for consistency with core-back

File: di-ipv-cimit-stub/deploy/template.yaml (hunk 6 above)
Comment (rpayal) [reply di-ipv-cimit-stub/deploy/template.yaml]: done

File: di-ipv-cimit-stub/deploy/template.yaml (hunk 7 above)
Comment (rpayal) [reply di-ipv-cimit-stub/deploy/template.yaml]: done

File: di-ipv-cimit-stub/deploy/template.yaml (hunk 8 above)
Comment (rpayal) [reply di-ipv-cimit-stub/deploy/template.yaml]: done

File: di-ipv-cimit-stub/deploy/template.yaml (hunk 16 above)
Comment (rpayal) [reply di-ipv-cimit-stub/deploy/template.yaml]: done

File: di-ipv-cimit-stub/deploy/template.yaml (hunk 17 above)
Comment (rpayal) [reply di-ipv-cimit-stub/deploy/template.yaml]: done

File: di-ipv-cimit-stub/gradle/wrapper/gradle-wrapper.jar
Comment (rpayal) [reply di-ipv-cimit-stub/gradle/wrapper/gradle-wrapper.jar:1]: Changed as in core - gradle-7.3-bin.zip

**Review by therealslimsladey — APPROVED** [review]
